Support 32 bit RGBA sprite sequences

pchote opened the pull request:
Last year, #15194 unified Indexed vs RGBA sprites at the shader level and removed the performance cliff caused by switching between them. This PR takes another step on the slow-burning renderer rewrite by carrying the unification through the `Sequence` and `SpriteLoader` code.

Modders should not get too excited just yet: all the caveats of not having a palette still apply, so RGBA sprites won't yet work with player colours, target flashes, map lighting, alpha/shadows, and probably more. I expect that this won't be properly usable until we implement vertex-based color tinting/alpha, mask or chroma-key based player colors, and #3692.

The testcase demonstrates @mabulsoud's radar dome from #13245 ingame.

A more in-depth testcase is available on my [d2k-r16](https://github.com/pchote/OpenRA/tree/d2k-r16) branch, which switches D2k to use the higher quality R16 sprites (which you will need to find and install manually).

pchote committed:
Support 32 bit png sprites.

dragunoff commented:
I've been using this PR for a while and successfully loading 32-bit sprites but I noticed that R and B channels are swapped. If I for example pass this in my sprite's image data: `(uint)Color.Red.ToArgb()` it would render as blue.

dragunoff commented on OpenRA.Mods.Common/SpriteLoaders/PngSheetLoader.cs:
```diff
@@ -69,24 +66,29 @@ public bool TryParseSprite(Stream s, out ISpriteFrame[] frames, out TypeDictiona
 			{
 				var frameStart = frameRegions[i].X + frameRegions[i].Y * png.Width;
```
I think these should be multiplied by `pixelLength` to get the offset for the frame:
```csharp
var frameStart = frameRegions[i].X * pixelLength + frameRegions[i].Y * pixelLength * png.Width;
```

dragunoff commented:
Another weird thing I noticed is that some images are cut off (missing one row or column of pixels), the TD icon for example:
![изображение](https://user-images.githubusercontent.com/1355810/64211380-a89cf200-ceae-11e9-83dc-efd39e272100.png)

abcdefg30 commented:
Are you certain this is a regression from this PR? We had issues with such missing lines before as well.

dragunoff commented:
> 
> 
> Are you certain this is a regression from this PR? We had issues with such missing lines before as well.

No, I'm not sure about that. But I think it tends to happen with 32-bit sprites.

pchote commented:
@dragunoff can you provide your testcases?

dragunoff commented:
@pchote Please have a look at https://github.com/dragunoff/OpenRA/commit/87131231388884d86a7b3e689c66935b602fb71c

I've added some simple color bars that render fine a single frame but are all jumbled up when sliced into frames. I have not tested the sprites in-game, only in the asset browser. I hope the metadata in the frames is correct. The TD logo also renders with the first row of the image missing (screenshot in an earlier message).

dragunoff commented:
The problem I illustrated above is actually caused by the wrong frame offset calculation that I already mentioned in https://github.com/OpenRA/OpenRA/pull/16885#discussion_r320494401.

After fixing that the problem goes away and frames render correctly (see 
https://github.com/dragunoff/OpenRA/commit/ad7f468d486d4cfba7959719dd5b7e8d4fd543ae). 

I will check again how sprites render on my downstream branch and if I am still able to reproduce the problem.

dragunoff commented:
OK, this is happening **only on Windows and only in the asset browser**. It also affects 8-bit PNGs and my custom sprite loaders so it's probably a different issue altogether (something with the asset browser maybe?). I discovered this by testing on an Ubuntu machine where everything looks fine both in-game and in the asset browser. Here's what I see in the asset browser on Windows 7:
![vQxkBqADr0](https://user-images.githubusercontent.com/1355810/71217852-dc6a1f00-22c7-11ea-862b-cc1d12520867.gif)

pchote commented:
Rebased and fixed the frame offset issue. I still need to look into the reversed color channels.

pchote commented:
Updated. The color channel issue was a silly oversight: OpenRA internally uses BGRA channel order, not RGBA. Renamed `SpriteFrameType.RGBA` to `SpriteFrameType.BGRA` to account for this.

The cursor changes turned into a bigger rework, so I will file these shortly as a followup PR.

Would you mind giving this another look @dragunoff?

pchote committed:
Add RGBA cursor support.

dragunoff commented:
@pchote Seems fine on my downstream branch and the testcases that I posted earlier look good too. The asset browser glitch on Windows 7 is still there.

pchote commented:
Do you have Windows configured with >100% DPI scaling and `Graphics.DisableWindowsDPIScaling` disabled in your settings.yaml? This could easily cause both the asset browser glitch here and the cursor distortion in #17517 without the patches on my `render-wip` branch.

dragunoff commented:
@pchote How do I check for these settings?

pchote commented:
Open `settings.yaml` in your support dir and delete the `DisableWindowsDPIScaling` line if it exists under the `Graphics` section.

dragunoff commented:
No such line there and I don't think I have messed with the DPI scaling in Windows:
![изображение](https://user-images.githubusercontent.com/1355810/71563363-e8f43180-2a96-11ea-8cc0-7475aa64b833.png)

dragunoff commented:
@pchote Does this PR add support for 32-bit sprite annotations? I'm talking about the move flash in particular. I tried but it doesn't seem to work.

dragunoff commented:
@pchote Please ignore my previous comment - move flash works beatifuly, I had a mistake in the yaml.

pchote committed:
Use 32bit modcontent cursor.

pchote commented:
Rebased on #17517 and added the (now much smaller) RGBA cursor commits here.

I have not yet fixed the color flipping issue (caused by incorrectly caching data) raised by @dragunoff in #17517.

pchote commented:
Updated. Duplicated cursors should now work correctly.

I can't reproduce the asset browser png issue in my Windows 10 VM, so I don't know what is going on there.

pchote commented:
Rebased.

pchote commented:
I would really like to remove the requirement to restart after changing the "Increase Command Cursor Size" setting for #17345, but doing this cleanly requires merging `HardwareCursor` and `SoftwareCursor`, which requires this PR.

There should be nothing controversial, or even particularly complicated here. Can we please try and make some progress here?

abcdefg30 commented on OpenRA.Game/Graphics/Util.cs:
```diff
@@ -61,23 +61,57 @@ public static void FastCreateQuad(Vertex[] vertices, float3 a, float3 b, float3
+			}
+			else
+			{
+				var data = dest.Sheet.GetData();
```
This seems to be redundant to `destData`?

abcdefg30 commented on OpenRA.Mods.Common/UtilityCommands/DumpSequenceSheetsCommand.cs:
```diff
@@ -44,14 +44,19 @@ void IUtilityCommand.Run(Utility utility, string[] args)
 
+			sb = sequences.SpriteCache.SheetBuilders[SpriteFrameType.BGRA];
+			foreach (var s in sb.AllSheets)
+					s.AsPng().Save("{0}.png".F(count++));
```
The indentation here looks wrong.

reaperrr commented on OpenRA.Mods.D2k/SpriteLoaders/R8Loader.cs:
```diff
@@ -22,6 +22,8 @@ public class R8Loader : ISpriteLoader
 		class R8Frame : ISpriteFrame
 		{
+			public SpriteFrameType Type { get; set; }
+
```
Minor consistency nit (not a blocker): This is the only loader you added an empty line to. Might want to drop this one while adressing @abcdefg30's points.

reaperrr reviewed (dismissed):
Ingame looks good, didn't spot any code issues.

pchote commented:
Fixed, rebased, and removed testcase.

abcdefg30 commented on OpenRA.Game/Graphics/Util.cs:
```diff
@@ -1,6 +1,6 @@
 #region Copyright & License Information
 /*
- * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * Copyright 2007-2019 The OpenRA Developers (see AUTHORS)
```
Time travel.

pchote committed:
Add support for 32 bit BGRA sprites.

pchote commented:
Fixed.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/0da929264049105ad228082970171d2bec475f43...a228e94ffcd53ab4be155b5b86f160916761a0e6)
